### ISSUE TITLE
Compiler complaining about dead code when enum is used in a match against return value of FFI call

### ISSUE THREAD

**Issue by felipeamp** [rust-lang/rust#85677]
I tried this code:

```rust
#[repr(C)]
#[derive(Debug, PartialEq)]
enum ResultEnumFromC {
    Success = 0,
    FailureOfTypeOne = 1,
    FailureOfTypeTwo = 2,
}

pub enum ErrEnumInRust {
    FailureOfTypeOne = 0,
    FailureOfTypeTwo = 1,
}

extern "C" {
    fn FunctionFromC() -> ResultEnumFromC;
}

pub fn wrapper_around_function_from_c(
) -> Result<(), ErrEnumInRust> {
    match unsafe { FunctionFromC() } {
        ResultEnumFromC::Success => Ok(()),
        ResultEnumFromC::FailureOfTypeOne => Err(ErrEnumInRust::FailureOfTypeOne),
        ResultEnumFromC::FailureOfTypeTwo => Err(ErrEnumInRust::FailureOfTypeTwo),
    }
}
```
link to example in Rust playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c3d9e0b917bfabeebeb6285a4aa73a6d

I expected to see this happen: no warning about dead code, since ResultEnumFromC is created on the C side and used in a match on the Rust side.

Instead, this happened: Compiler complained about ResultEnumFromC never being constructed (which is true on the Rust side, but I wouldn't expect this since it's the result of an FFI call).

```
warning: variant is never constructed: `Success`
 --> src/main.rs:4:5
  |
4 |     Success = 0,
  |     ^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: variant is never constructed: `FailureOfTypeOne`
 --> src/main.rs:5:5
  |
5 |     FailureOfTypeOne = 1,
  |     ^^^^^^^^^^^^^^^^^^^^

warning: variant is never constructed: `FailureOfTypeTwo`
 --> src/main.rs:6:5
  |
6 |     FailureOfTypeTwo = 2,
  |     ^^^^^^^^^^^^^^^^^^^^

warning: 3 warnings emitted
```

### Meta

Happens on Rust playground on stable, beta and nightly.

`rustc --version --verbose`:
```
1.52.1
```

<details><summary>Backtrace</summary>
<p>

```
warning: variant is never constructed: `Success`
 --> src/main.rs:4:5
  |
4 |     Success = 0,
  |     ^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: variant is never constructed: `FailureOfTypeOne`
 --> src/main.rs:5:5
  |
5 |     FailureOfTypeOne = 1,
  |     ^^^^^^^^^^^^^^^^^^^^

warning: variant is never constructed: `FailureOfTypeTwo`
 --> src/main.rs:6:5
  |
6 |     FailureOfTypeTwo = 2,
  |     ^^^^^^^^^^^^^^^^^^^^

warning: 3 warnings emitted
```

</p>
</details>

**Comment by WaffleLapkin** [comment]
MRE:
```rust
extern {
    fn f() -> E;
}

#[repr(C)]
enum E {
    V
}

pub fn g() {
    let e = unsafe { f() };
    match e {
        E::V => ()
    }
}
```
```
warning: variant `V` is never constructed
 --> src/lib.rs:7:5
  |
6 | enum E {
  |      - variant in this enum
7 |     V
  |     ^
  |
  = note: `#[warn(dead_code)]` on by default
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b49886b8403eec92e01d75a500734206))

**Comment by klensy** [comment]
Found that in #127274 with `TypeKind` enum.
